cpp/src/gandiva/precompiled/decimal_wrapper.cc

```diff
@@ -231,4 +231,131 @@ void castDECIMAL_decimal128_internal(int64_t x_high, uint64_t x_low, int32_t x_p
   *out_low = out.low_bits();
 }
 
+FORCE_INLINE
+int32_t hash32_decimal128_internal(int64_t x_high, uint64_t x_low, int32_t x_precision,
+                                   int32_t x_scale, boolean x_validity) {
+  return x_validity
+             ? hash32_buf(gandiva::BasicDecimal128(x_high, x_low).ToBytes().data(), 16, 0)
+             : 0;
+}
+
+FORCE_INLINE
+int32_t hash_decimal128_internal(int64_t x_high, uint64_t x_low, int32_t x_precision,
+                                 int32_t x_scale, boolean x_validity) {
+  return hash32_decimal128_internal(x_high, x_low, x_precision, x_scale, x_validity);
+}
+
+FORCE_INLINE
+int64_t hash64_decimal128_internal(int64_t x_high, uint64_t x_low, int32_t x_precision,
+                                   int32_t x_scale, boolean x_validity) {
+  return x_validity
+             ? hash64_buf(gandiva::BasicDecimal128(x_high, x_low).ToBytes().data(), 16, 0)
+             : 0;
+}
+
+FORCE_INLINE
+int32_t hash32WithSeed_decimal128_internal(int64_t x_high, uint64_t x_low,
+                                           int32_t x_precision, int32_t x_scale,
+                                           boolean x_validity, int32_t seed,
+                                           boolean seed_validity) {
+  return x_validity && seed_validity
+             ? hash32_buf(gandiva::BasicDecimal128(x_high, x_low).ToBytes().data(), 16,
+                          seed)
+             : 0;
```

Comment:
i am not sure this logic is right
if input is not valid, we should return seed as the hash
if seed is not valid, we should hash value with 0 as seed instead of 0 hash i guess..
Lets discuss with Ravindra on Monday.